Make HitRenderer not have a publicly-settable Beatmap, replace AllObj…

smoogipoo opened the pull request:
Just a bit of tidying up for now.

smoogipoo committed:
Make HitRenderer not have a publicly-settable Beatmap, replace AllObjectsJudged for performance.

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into refactor_hitrenderer

Conflicts:
	osu.Game.Modes.Catch/CatchRuleset.cs
	osu.Game.Modes.Mania/ManiaRuleset.cs
	osu.Game.Modes.Osu/OsuRuleset.cs
	osu.Game.Modes.Taiko/TaikoRuleset.cs

peppy commented on osu.Game.Modes.Mania/UI/ManiaHitRenderer.cs:
```diff
@@ -12,7 +13,8 @@ public class ManiaHitRenderer : HitRenderer<ManiaBaseHit>
         private readonly int columns;
 
-        public ManiaHitRenderer(int columns = 5)
+        public ManiaHitRenderer(Beatmap beatmap, int columns = 5)
```
i feel like column count won't be required as it should be part of the beatmap data here. but we can address this when mania is actually implemented. (although it does bring the question to mind of whether we should be passing in a `WorkingBeatmap` instead of just `Beatmap`, so we have mods etc.).

smoogipooo replied:
Definitely re: when mania is implemented.

Actually very interesting point re: workingbeatmap, because then the hitrenderer could also do Apply/Remove for each mod on the hitobjects internally, along with perhaps moving the current auto mod application inside too.

peppy commented on osu.Game/Modes/UI/HitRenderer.cs:
```diff
@@ -28,46 +26,63 @@ public abstract class HitRenderer : Container
-            }
-        }
+        protected abstract HitObjectConverter<TObject> Converter { get; }
+        protected virtual List<TObject> Convert(Beatmap beatmap) => Converter.Convert(beatmap);
```
why does this still take a beatmap parameter? also this can probably be made private or even inlined in its single usage.

smoogipooo replied:
In the end the converter isn't going to exist exactly as it is now, but I haven't worked on it yet. This is just a precursory branch to make the code here a bit more readable. Converter is my next task.

smoogipoo committed:
Move back to using an abstract method to determine if all objects have been judged.

Because sliderticks provide judgements even though they are added as nested hitobjects, the count method would not work to determine if all hitobjects have been judged. This needs a little bit more thought put in...

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into refactor_hitrenderer
